File: src/main/java/org/mitallast/queue/common/events/EventBus.java
package org.mitallast.queue.common.events;

import java.util.function.Consumer;

public interface EventBus {

    <Event> void subscribe(Class<Event> eventClass, Consumer<Event> consumer);

    <Event> void unsubscribe(Class<Event> eventClass, Consumer<Event> consumer);

    <Event> void trigger(Event event);

    <Event> void trigger(Class<Event> eventClass, Event event);
}

File: src/main/java/org/mitallast/queue/common/events/EventBus.kt
package org.mitallast.queue.common.events

import java.util.concurrent.Executor

interface EventBus {

    fun <Event : Any> subscribe(eventClass: Class<Event>, consumer: (Event) -> Unit)

    fun <Event : Any> subscribe(eventClass: Class<Event>, consumer: (Event) -> Unit, executor: Executor)

    fun <Event : Any> unsubscribe(eventClass: Class<Event>, consumer: (Event) -> Unit)

    fun <Event : Any> trigger(event: Event)
}
